src/platform/backends/qemu/qemu_virtual_machine.cpp

```diff
@@ -261,6 +261,8 @@ void mp::QemuVirtualMachine::start()
     }
 
     vm_process->start();
+    state = VirtualMachine::State::starting;
```

Comment:
Why did you need to move this here? We should be holding the new lock at this point and `on_started` would be called before returning from the next line, because it's [registered as a callback for when the process starts](https://github.com/canonical/multipass/blob/4b03174758438540a0b6d3ec2851c446e797d185/src/platform/backends/qemu/qemu_virtual_machine.cpp#L474). Unless it fails to start, in which case it should actually not be in the `starting` state, right?

Reply:
I lock itself was not able to solve the issue because the second `start()` call was happening before the callback `on_started()` was able to be called by the process created by the first `start()` call.

Reply:
Hmm, I still don't get it. Can you offer an example of that? A stack trace would do (real or imaginary). I am probably missing something, but here's what I am seeing.

Until `on_started` is called, the main thread is still [waiting for it to start](https://github.com/canonical/multipass/pull/2711/files#diff-b70c5ab44722c02494a880259a18034fce5b14fb187dcc9830192a02d50078d4R266), and so still holding the lock. Notice that [`waitForStarted`](https://doc.qt.io/qt-5/qprocess.html#waitForStarted) *Blocks until the process has started and the `started()` signal has been emitted*. 

The `started()` signal is connected in the code I linked above with a [`DirectConnection`](https://doc.qt.io/qt-5/qt.html#ConnectionType-enum) (that's what `AutoConnection` means in this case since the receiving lambda is created when connecting). So, if the signal hasn't been handled, it hasn't been emitted. And we just saw that, if `started()` hasn't been emitted, `waitForStarted` hasn't returned yet (or there was an error).

Reply:
So, after re-testing this again it seems like the lock in of itself is enough to prevent concurrent `start()` commands. And, I would tend to agree with you in that `waitForStarted()` should block until the process has started, but thought that changing where the vm state is updated was necessary.